[funcol] a few optimizations to funcol

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113324

Apply a few optimizations to funcol:

- allgather on non-0 dim, the resulting tensor already needs to access
data in order to do torch.cat, so we sync wait here so that we don;t
need to go through ACT dispatch for chunk + cat alltogether
- have a fast return logic to aten.view as it's a commonly hit op for
view related ops

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @lucasllc @d4l3k

wanchaol committed:
[funcol] a few optimizations to funcol

Apply a few optimizations to funcol:

- allgather on non-0 dim, the resulting tensor already needs to access
data in order to do torch.cat, so we sync wait here so that we don;t
need to go through ACT dispatch for chunk + cat alltogether
- have a fast return logic to aten.view as it's a commonly hit op for
view related ops

[ghstack-poisoned]

wanchaol committed:
Update on "[funcol] a few optimizations to funcol"


Apply a few optimizations to funcol:

- allgather on non-0 dim, the resulting tensor already needs to access
data in order to do torch.cat, so we sync wait here so that we don;t
need to go through ACT dispatch for chunk + cat alltogether
- have a fast return logic to aten.view as it's a commonly hit op for
view related ops

[ghstack-poisoned]

wanchaol committed:
Update on "[funcol] a few optimizations to funcol"


Apply a few optimizations to funcol:

- allgather on non-0 dim, the resulting tensor already needs to access
data in order to do torch.cat, so we sync wait here so that we don;t
need to go through ACT dispatch for chunk + cat alltogether
- have a fast return logic to aten.view as it's a commonly hit op for
view related ops

[ghstack-poisoned]

fegin commented on torch/distributed/_functional_collectives.py:
```diff
@@ -192,6 +192,9 @@ def all_gather_tensor(
     if gather_dim != 0:
+        # torch.cat access the data so we already need to wait here, first do wait
+        # and then chunk + cat avoid us going through ACT dispatching logic again
+        res = res.wait()  # type: ignore[attr-defined]
```
iirc, this is only a performance optimization. No logic is changed.

wanchaol replied:
Yep.

fegin commented on torch/distributed/_functional_collectives.py:
```diff
@@ -412,6 +415,14 @@ def _get_acs_underlying_tensor(self):
 
     @classmethod
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        if func == torch.ops.aten.wait.default:
```
`torch.ops.aten.view.default`?

wanchaol replied:
err, good catch!!

wanchaol committed:
Update on "[funcol] a few optimizations to funcol"


Apply a few optimizations to funcol:

- allgather on non-0 dim, the resulting tensor already needs to access
data in order to do torch.cat, so we sync wait here so that we don;t
need to go through ACT dispatch for chunk + cat alltogether
- have a fast return logic to aten.view as it's a commonly hit op for
view related ops

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

XilunWu reviewed (approved):
I think this PR is ready to merge?

wanchaol committed:
Update on "[funcol] a few optimizations to funcol"


Apply a few optimizations to funcol:

- allgather on non-0 dim, the resulting tensor already needs to access
data in order to do torch.cat, so we sync wait here so that we don;t
need to go through ACT dispatch for chunk + cat alltogether
- have a fast return logic to aten.view as it's a commonly hit op for
view related ops

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change here, but it is failing internal test' -c ghfirst

Here is the failed SC job https://www.internalfb.com/sandcastle/workflow/603482350069087187, it has something to do with `'FakeTensor' object has no attribute 'wait'`:

```
----------------------------------------------------------------------
Ran 1 test in 24.547s

FAILED (errors=1)
[2023-11-15 13:33:43,141] torch._dynamo.utils: [INFO] TorchDynamo compilation metrics:
[2023-11-15 13:33:43,141] torch._dynamo.utils: [INFO] Function    Runtimes (s)
[2023-11-15 13:33:43,141] torch._dynamo.utils: [INFO] ----------  --------------

RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/fx/passes/infra/pass_manager.py", line 270, in __call__
    res = fn(module)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/fx/passes/infra/pass_base.py", line 40, in __call__
    res = self.call(graph_module)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/distributed/_tensor/experimental/tp_transform.py", line 94, in call
    _partitioner(gm)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/distributed/_tensor/experimental/tp_transform.py", line 371, in _partitioner
    _insert_reshard_gm(
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/distributed/_tensor/experimental/tp_transform.py", line 449, in _insert_reshard_gm
    reshard_gm = make_fx(reshard_fn)(input_arg_tensor)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/fx/experimental/proxy_tensor.py", line 863, in wrapped
    t = dispatch_trace(wrap_key(func, args, fx_tracer, pre_dispatch), tracer=fx_tracer, concrete_args=tuple(phs))
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/_compile.py", line 24, in inner
    return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/_dynamo/eval_frame.py", line 408, in _fn
    return fn(*args, **kwargs)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/fx/experimental/proxy_tensor.py", line 481, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/_dynamo/eval_frame.py", line 408, in _fn
    return fn(*args, **kwargs)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/fx/_symbolic_trace.py", line 821, in trace
    (self.create_arg(fn(*args)),),
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/fx/experimental/proxy_tensor.py", line 517, in wrapped
    out = f(*tensors)
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/distributed/_tensor/experimental/tp_transform.py", line 443, in reshard_fn
    return redistribute_local_tensor(
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/distributed/_tensor/redistribute.py", line 123, in redistribute_local_tensor
    new_local_tensor = current_placement._to_replicate_tensor(
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/distributed/_tensor/placement_types.py", line 247, in _to_replicate_tensor
    result = funcol.all_gather_tensor(
  File "/data/users/huydo/fbsource/buck-out/v2/gen/fbcode/573348b686d4158f/caffe2/test/distributed/_tensor/experimental/__tp_transform__/tp_transform#link-tree/torch/distributed/_functional_collectives.py", line 197, in all_gather_tensor
    res = res.wait()  # type: ignore[attr-defined]
AttributeError: 'FakeTensor' object has no attribute 'wait'
```

huydhn commented:
I wonder why OSS tests doesn't catch this

fduwjj commented on torch/distributed/_functional_collectives.py:
```diff
@@ -412,6 +415,14 @@ def _get_acs_underlying_tensor(self):
 
     @classmethod
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        if func == torch.ops.aten.view.default:
```
I am ok with perf improvement change, but would this make the code look hacky?

XilunWu commented:
Adding `wait()` to `FakeTensor` should fix the issue. Is there anything else we miss except this?

wanchaol committed:
Update on "[funcol] a few optimizations to funcol"


Apply a few optimizations to funcol:

- allgather on non-0 dim, the resulting tensor already needs to access
data in order to do torch.cat, so we sync wait here so that we don;t
need to go through ACT dispatch for chunk + cat alltogether
- have a fast return logic to aten.view as it's a commonly hit op for
view related ops

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

wanchaol commented:
@pytorchbot rebase

wanchaol committed:
Update on "[funcol] a few optimizations to funcol"


Apply a few optimizations to funcol:

- allgather on non-0 dim, the resulting tensor already needs to access
data in order to do torch.cat, so we sync wait here so that we don;t
need to go through ACT dispatch for chunk + cat alltogether
- have a fast return logic to aten.view as it's a commonly hit op for
view related ops

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]
